agent/xds/routes.go

```diff
@@ -37,55 +37,101 @@ func routesFromSnapshotConnectProxy(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.M
-}
-
-func routesFromUpstreams(snap proxycfg.ConfigSnapshotUpstreams, upstreams structs.Upstreams) ([]proto.Message, error) {
 	var resources []proto.Message
-
-	for _, u := range upstreams {
+	for _, u := range cfgSnap.Proxy.Upstreams {
 		upstreamID := u.Identifier()
 
 		var chain *structs.CompiledDiscoveryChain
 		if u.DestinationType != structs.UpstreamDestTypePreparedQuery {
-			chain = snap.DiscoveryChain[upstreamID]
+			chain = cfgSnap.ConnectProxy.DiscoveryChain[upstreamID]
 		}
 
 		if chain == nil || chain.IsDefault() {
 			// TODO(rb): make this do the old school stuff too
 		} else {
-			upstreamRoute, err := makeUpstreamRouteForDiscoveryChain(&u, chain)
+			virtualHost, err := makeUpstreamRouteForDiscoveryChain(upstreamID, chain, []string{"*"})
 			if err != nil {
 				return nil, err
 			}
-			if upstreamRoute != nil {
-				resources = append(resources, upstreamRoute)
+
+			route := &envoy.RouteConfiguration{
+				Name:         upstreamID,
+				VirtualHosts: []envoyroute.VirtualHost{virtualHost},
+				// ValidateClusters defaults to true when defined statically and false
+				// when done via RDS. Re-set the sane value of true to prevent
+				// null-routing traffic.
+				ValidateClusters: makeBoolValue(true),
 			}
+			resources = append(resources, route)
 		}
 	}
 
 	// TODO(rb): make sure we don't generate an empty result
-
 	return resources, nil
 }
 
+// routesFromSnapshotIngressGateway returns the xDS API representation of the
+// "routes" in the snapshot.
+func routesFromSnapshotIngressGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+	if cfgSnap == nil {
+		return nil, errors.New("nil config given")
+	}
+
+	var result []proto.Message
+	for listenerKey, upstreams := range cfgSnap.IngressGateway.Upstreams {
+		// Do not create any route configuration for TCP listeners
+		if listenerKey.Protocol == "tcp" {
+			continue
+		}
+
+		upstreamRoute := &envoy.RouteConfiguration{
+			Name: listenerKey.RouteName(),
+			// ValidateClusters defaults to true when defined statically and false
+			// when done via RDS. Re-set the sane value of true to prevent
+			// null-routing traffic.
+			ValidateClusters: makeBoolValue(true),
+		}
+		for _, u := range upstreams {
+			upstreamID := u.Identifier()
+			chain := cfgSnap.IngressGateway.DiscoveryChain[upstreamID]
+			if chain == nil {
+				continue
+			}
+
+			var domains []string
+			switch {
+			case len(upstreams) == 1:
+				// Don't require a service prefix on the domain if there is only 1
+				// upstream. This makes it a smoother experience when only having a
+				// single service associated to a listener, which is probably a common
+				// case when demoing/testing
+				domains = []string{"*"}
+			case len(u.IngressHosts) > 0:
+				// If a user has specified hosts, do not add the default
+				// "<service-name>.*" prefix
+				domains = u.IngressHosts
+			default:
+				domains = []string{fmt.Sprintf("%s.*", chain.ServiceName)}
```

Comment:
Should this be `<service name>.<namespace>.*` in enterprise? Or maybe `[]string{"<service name>.<namespace>.*", "<service name>.*}` in the case of only a single namespace being used for all services for this listener. 

Should we support a single listener with services for multiple namespaces?

Reply:
Hmm, just realized that the DNS name we provide is `<service-name>.ingress.<namespace>.<datacenter>.<domain>`, so to accurately match our own domains we would need to add the `.ingress` label in here. 🤔 

In terms of supporting a single listener with services for multiple namespaces, I think as long as we can support it in a coherent way, that allowing that flexibility to the user is a plus. I can't think of anything inherently wrong about allowing the services to go over the same listener. If a user does not want them to be on the same listener, they can always arrange the config entry in that way.